### PR TITLE
Event generation to report network statistics

### PR DESCRIPTION
**Pull request by ptrivedi**
Upon data acknowledgement, indicate a connection event with updated RTT, congestion window, unacknowledged bytes etc. values

## Description

_Describe the purpose of and changes within this Pull Request._

## Testing

_Do any existing tests cover this change? Are new tests needed?_

## Documentation

_Is there any documentation impact for this change?_

### CODE REVIEW

File: src/inc/msquic.h [hunk 1]
```diff
@@ -1161,6 +1161,7 @@ typedef enum QUIC_CONNECTION_EVENT_TYPE {
 #ifdef QUIC_API_ENABLE_PREVIEW_FEATURES
     QUIC_CONNECTION_EVENT_RELIABLE_RESET_NEGOTIATED         = 16,   // Only indicated if QUIC_SETTINGS.ReliableResetEnabled is TRUE.
     QUIC_CONNECTION_EVENT_ONE_WAY_DELAY_NEGOTIATED          = 17,   // Only indicated if QUIC_SETTINGS.OneWayDelayEnabled is TRUE.
+    QUIC_CONNECTION_EVENT_DATA_ACKED                        = 18,
```
Comment (nibanks) [inline src/inc/msquic.h:1164]: I wouldn't call this `DATA_ACKED`. Maybe `NETWORK_STATISTICS`

Comment (ptrivedi) [reply src/inc/msquic.h:1164]: Well, the event triggers upon data acked. What's being returned is net stats. It felt a little strange to call it a 'network stats' event. Do we anticipate someone using this event as a pure indication of data being acknowledged to do something other than network stats with it?

Comment (nibanks) [reply src/inc/msquic.h:1164]: We might end up triggered for other things too

Comment (ptrivedi) [reply src/inc/msquic.h]: Ok fair, can change.

File: src/inc/msquic.h [hunk 2]
```diff
@@ -1161,6 +1161,7 @@ typedef enum QUIC_CONNECTION_EVENT_TYPE {
 #ifdef QUIC_API_ENABLE_PREVIEW_FEATURES
     QUIC_CONNECTION_EVENT_RELIABLE_RESET_NEGOTIATED         = 16,   // Only indicated if QUIC_SETTINGS.ReliableResetEnabled is TRUE.
     QUIC_CONNECTION_EVENT_ONE_WAY_DELAY_NEGOTIATED          = 17,   // Only indicated if QUIC_SETTINGS.OneWayDelayEnabled is TRUE.
+    QUIC_CONNECTION_EVENT_NETWORK_STATISTICS                = 18,
```
Comment (nibanks) [inline src/inc/msquic.h:1164]: The new event must be disabled by default, and opt in. Additionally, we'll need to a test to validate this behavior.

File: docs/api/QUIC_SETTINGS.md [hunk 3]
```diff
@@ -53,6 +53,7 @@ typedef struct QUIC_SETTINGS {
             uint64_t EncryptionOffloadAllowed               : 1;
             uint64_t ReliableResetEnabled                   : 1;
             uint64_t OneWayDelayEnabled                     : 1;
+            uint64_t NetStatsEventEnabled                   : 1;
             uint64_t RESERVED                               : 23;
```
Comment (nibanks) [inline docs/api/QUIC_SETTINGS.md:57]:
```suggestion
            uint64_t RESERVED                               : 22;
```

File: docs/api/QUIC_SETTINGS.md [hunk 4]
```diff
@@ -102,6 +103,7 @@ typedef struct QUIC_SETTINGS {
             uint64_t EncryptionOffloadAllowed  : 1;
             uint64_t ReliableResetEnabled      : 1;
             uint64_t OneWayDelayEnabled        : 1;
+            uint64_t NetStatsEventEnabled      : 1;
             uint64_t ReservedFlags             : 60;
```
Comment (nibanks) [inline docs/api/QUIC_SETTINGS.md:107]:
```suggestion
            uint64_t ReservedFlags             : 59;
```

File: src/core/unittest/SettingsTest.cpp [hunk 5]
```diff
@@ -76,7 +76,7 @@ static uint32_t PopCount(T Value) {
     return Count;
 }
 
-TEST(SettingsTest, TestAllSettingsFieldsSet)
+/* TEST(SettingsTest, TestAllSettingsFieldsSet)
```
Comment (nibanks) [inline src/core/unittest/SettingsTest.cpp:79]: Please reenable. I think you just need to fix the reserved fields like I pointed out above.

File: src/core/quicdef.h [hunk 6]
```diff
@@ -532,6 +532,11 @@ CXPLAT_STATIC_ASSERT(
 //
 #define QUIC_DEFAULT_ONE_WAY_DELAY_ENABLED           FALSE
 
+//
+// The default settings for allowing Network Statistics event to be raised.
+//
+#define QUIC_DEFAULT_NET_STATS_EVENT_ENABLED           FALSE
```
Comment (nibanks) [inline src/core/quicdef.h:538]: nit: can you align the `FALSE` with (most) of the rest above.

File: src/core/cubic.c [hunk 7]
```diff
@@ -669,6 +669,33 @@ CubicCongestionControlOnDataAcknowledged(
 
     Cubic->TimeOfLastAck = TimeNowUs;
     Cubic->TimeOfLastAckValid = TRUE;
+
+#if defined(QUIC_API_ENABLE_PREVIEW_FEATURES)
```
Comment (nibanks) [inline src/core/cubic.c:673]: You don't wrap the product code with this (we always define it internally).

File: src/core/bbr.c [hunk 8]
```diff
@@ -739,6 +770,11 @@ BbrCongestionControlOnDataAcknowledged(
         BbrCongestionControlUpdateCongestionWindow(
             Cc, AckEvent->NumTotalAckedRetransmittableBytes, AckEvent->NumRetransmittableBytes);
 
+#if defined(QUIC_API_ENABLE_PREVIEW_FEATURES)
```
Comment (nibanks) [inline src/core/bbr.c:773]: Remove these from the product code.

File: src/core/settings.c [hunk 9]
```diff
@@ -1405,6 +1425,7 @@ QuicSettingsDump(
     QuicTraceLogVerbose(SettingEncryptionOffloadAllowed,    "[sett] EncryptionOffloadAllowed = %hhu", Settings->EncryptionOffloadAllowed);
     QuicTraceLogVerbose(SettingReliableResetEnabled,        "[sett] ReliableResetEnabled   = %hhu", Settings->ReliableResetEnabled);
     QuicTraceLogVerbose(SettingOneWayDelayEnabled,          "[sett] OneWayDelayEnabled     = %hhu", Settings->OneWayDelayEnabled);
+    QuicTraceLogVerbose(SettingNetStatsEventEnabled,        "[sett] NetStatsEventEnabled   = %hhu", Settings->NetStatsEventEnabled);
```
Comment (nibanks) [inline src/core/settings.c:1428]: You need to run `.\scripts\update-sidecar.ps1`

File: src/inc/msquic.h [hunk 10]
```diff
@@ -1243,6 +1246,14 @@ typedef struct QUIC_CONNECTION_EVENT {
...
             BOOLEAN ReceiveNegotiated;          // TRUE if receiving one-way delay timestamps is negotiated.
         } ONE_WAY_DELAY_NEGOTIATED;
+        struct {
+           uint32_t BytesInFlight;              // Bytes that were sent on the wire, but not yet acked
+           uint64_t PostedBytes;                // Total bytes queued, but not yet acked. These may contain sent bytes that may have portentially lost too.
+           uint64_t IdealBytes;                 // Ideal number of bytes required to be available to  avoid limiting throughput
+           uint64_t SmoothedRTT;                // Smoothed RTT value
+           uint32_t CongestionWindow;           // Congestion Window
+           uint64_t Bandwidth;                  // Estimated bandwidth
+        } NETWORK_STATISTICS;
```
Comment (nibanks) [inline src/inc/msquic.h:1256]: You need to run `.\scripts\generate-dotnet.ps1`.

File: src/test/MsQuicTests.h [hunk 11]
```diff
@@ -811,6 +814,10 @@ typedef struct {
     QUIC_CTL_CODE(27, METHOD_BUFFERED, FILE_WRITE_DATA)
     // int - Family
 
+#define IOCTL_QUIC_RUN_VALIDATE_NET_STATS_CONN_EVENT \
+    QUIC_CTL_CODE(28, METHOD_BUFFERED, FILE_WRITE_DATA)
```
Comment (nibanks) [inline src/test/MsQuicTests.h:818]: Add the new IOCTL at the end of the file and use the next number available (not 28, as it's already used). Then update the MAX define at the end. Then you will need to update `QUIC_IOCTL_BUFFER_SIZES` in `control.cpp`.

File: src/test/bin/winkernel/control.cpp [hunk 12]
```diff
@@ -918,6 +918,13 @@ QuicTestCtlEvtIoDeviceControl(
         QuicTestCtlRun(QuicTestValidateConnectionEvents(Params->Test));
         break;
 
+#ifdef QUIC_API_ENABLE_PREVIEW_FEATURES
```
Comment (nibanks) [inline src/test/bin/winkernel/control.cpp:921]: Please put at the end of the switch.

File: src/test/lib/EventTest.cpp [hunk 13]
```diff
@@ -186,6 +190,17 @@ struct ConnValidator {
...
+	if (EventsExpectedAtleastOnce != nullptr) {
+            uint8_t Count = 0;
+            while (EventsExpectedAtleastOnce[Count] != nullptr) {
+                if (EventsExpectedAtleastOnce[Count]->Type == Event->Type) {
+                    EventsExpectedAtleastOnce[Count]->Success = true;
+                    return; 
+	        }
+	        Count++;
+	    }
+	}
```
Comment (nibanks) [inline src/test/lib/EventTest.cpp:202]: nit: please fix alignment with other code

File: src/test/lib/EventTest.cpp [hunk 14]
```diff
@@ -202,6 +217,18 @@ struct ConnValidator {
...
+        if (EventsExpectedAtleastOnce == nullptr) { return true; }
+        uint8_t Count = 0;
+        while (EventsExpectedAtleastOnce[Count] != nullptr) {
+            if (EventsExpectedAtleastOnce[Count]->Success == false) {
+                *EventIndex = Count;
+                return false; 
+	    }
+	    Count++;
+	}
+	return true;
```
Comment (nibanks) [inline src/test/lib/EventTest.cpp:230]: nit: please fix alignment

File: src/test/lib/EventTest.cpp [hunk 15]
```diff
@@ -557,6 +584,205 @@ void QuicTestValidateConnectionEvents(uint32_t Test)
...
+            new(std::nothrow) ConnEventValidator(QUIC_CONNECTION_EVENT_SHUTDOWN_COMPLETE),
+            nullptr
+        },
+        ServerConfiguration
+    );
+
+    Client.SetEventsExpectedAtleastOnce(
+        new(std::nothrow) ConnEventValidator* [2] {
+            new(std::nothrow) ConnEventValidator(QUIC_CONNECTION_EVENT_NETWORK_STATISTICS),
+	    nullptr
```
Comment (nibanks) [inline src/test/lib/EventTest.cpp:628]:
```suggestion
            new(std::nothrow) ConnEventValidator(QUIC_CONNECTION_EVENT_NETWORK_STATISTICS),
	        nullptr
```

File: src/test/lib/EventTest.cpp [hunk 16]
```diff
@@ -557,6 +584,205 @@ void QuicTestValidateConnectionEvents(uint32_t Test)
...
+        }
+    );
+
+    Server.SetEventsExpectedAtleastOnce(
+        new(std::nothrow) ConnEventValidator* [2] {
+            new(std::nothrow) ConnEventValidator(QUIC_CONNECTION_EVENT_NETWORK_STATISTICS),
+	    nullptr
```
Comment (nibanks) [inline src/test/lib/EventTest.cpp:635]:
```suggestion
            new(std::nothrow) ConnEventValidator(QUIC_CONNECTION_EVENT_NETWORK_STATISTICS),
	        nullptr
```
Please fix others too

File: src/core/unittest/SettingsTest.cpp (hunk 5 above)
Comment (ptrivedi) [reply src/core/unittest/SettingsTest.cpp:79]: Was an accidental leftover. Reserved flags are already fixed in this PR, were just not fixed in the doc

File: src/test/lib/EventTest.cpp [hunk 17]
```diff
@@ -557,6 +584,205 @@ void QuicTestValidateConnectionEvents(uint32_t Test)
...
+                QuicAddrGetFamily(&ServerLocalAddr.SockAddr)),
+            ServerLocalAddr.GetPort()));
+
+    TEST_TRUE(Client.Complete.WaitTimeout(2000));
+    TEST_TRUE(Server.Complete.WaitTimeout(1000));
+
+    uint8_t EventIndex = UINT8_MAX;
+    if (!Server.SuccessOfEventsExpectedAtleastOnce(&EventIndex) && EventIndex != UINT8_MAX) {
+        TEST_FAILURE("QuicTestValidateNetStatsEvent1: Expected event did not occur. Event type %u  was expected to happen on server at least once", Server.EventsExpectedAtleastOnce[EventIndex]);
+    }
```
Comment (nibanks) [inline src/test/lib/EventTest.cpp:662]: I feel like a slightly better design is to push this logic all into `SuccessOfEventsExpectedAtleastOnce`. It loops through all events and if any were missed, you call `TEST_FAILURE`. Then here you only need:
```c
Server.SuccessOfEventsExpectedAtleastOnce();
```

Comment (ptrivedi) [reply src/test/lib/EventTest.cpp:662]: Def better. Done

**Review by nibanks — DISMISSED** [review]

**Review by nibanks — DISMISSED** [review]

**Review by nibanks — DISMISSED** [review]

**Review by nibanks — APPROVED** [review]

**Pull request merged**